advanced-multi-version-updates/v3/README.md

```diff
@@ -0,0 +1,165 @@
+# Guest Book Contract
+
+The [v2](../v2) contract was modified, removing the `payments` field and including that information in the `PostedMessage` structure.
+
+```rust
+pub struct PostedMessage {
+    pub payment: NearToken,
+    pub premium: bool,
+    pub sender: AccountId,
+    pub text: String,
+}
+
+pub struct GuestBook {
+  messages: Vector<PostedMessage>,
+  owner: AccountId
+}
+```
+
+If we deploy this contract on top of the [v2](../v2/) one and call any method we again will get the error:
+
+```
+panicked at 'Cannot deserialize the contract state.: ... }',
+```
+
+This is because the new contract expects to find `PostedMessages` with 4 fields (`payment`, `premium`, `sender`, `text`)
+but the saved messages only have 3 fields (they lack the `payment` field).
+
+In order to fix this problem we need to `migrate` the state, i.e. iterate through the current saved messages transforming them to the new version.
+
+We're getting rid of Vector of `payments` entirely, so just removing the field from the state won't cut it. The data is spread across multiple keys in the storage, so to clean it up properly, we need to explicitly call `payments::clear()`.
+
+```rust
+impl GuestBook {
+    // Upgrades from V2 to V3
+    fn unsafe_add_payment_to_message() {
+        let GuestBookV2 {
+            messages: old_messages,
+            mut payments,
+            owner,
+        } = env::state_read().unwrap();
+
+        let default_payment = NearToken::from_yoctonear(0);
+
+        // New messages must be written to storage
+        let mut messages = Vector::new(StorageKey::Messages);
+
+        for (idx, old_message) in old_messages.iter().enumerate() {
```

Comment:
I think this won't work. You have the same StorageKey in use. So they will be resolved in the  same storage key and you end up with both data (as you are pushing new in)

Reply:
I checked all migrations on [testnet](https://testnet.nearblocks.io/address/update-migrate-rust-advanced-multiversion-updates.testnet) - no issues found

New instance of `Vector` is created without knowledge of past data, and its length is set to zero in constructor. When a first message is added, it just overwrites the value at index zero and bumps the length.

Reply:
also, to prove it works, the contract I'm referencing to, returns a list of messages as follows (after V3 version was rolled out)
`near contract call-function as-read-only update-migrate-rust-advanced-multiversion-updates.testnet get_messages json-args {} network-config testnet now`

```
Result:
[
  {
    "payment": "100000000000000000000000",
    "premium": true,
    "sender": "denbite.testnet",
    "text": "a message"
  }
]
```

Reply:
Try to add an extra element in step 2 and you will face the error.
It looks like it's working but it is not. 
You have a Vector with 3 elements: 
```
|Element A|Element B|Element C|
```
                                                       
When you create a Vector::new(SameStorageKey)

You use the same memory but its index point is zero.
```
|Element A|Element B|Element C|
/\
|
```

So, you have two vectors. Both are pointing to the same place. When you iterate through the vector of 1 and push into the second, you replace Elements from the left.
It would probably be fine if your new element size is less or equal to the previous one. But it would fail otherwise.
But even in this case, you might face an issue that you still pay in the contract for space you don't use.
```
|Updated ELEMENT A IS BIGGER THAN PREVIOUS| ent B|Element C|
/\
|
```
After this push you will have a screwed next element. But it will still try to read the data and might even success. In the, end you will end up either with a failing migration or screwed data that is way worse :(

Reply:
Here is my V2 before migrating to V3:
```
[
  {
    "premium": false,
    "sender": "unnatural-error.testnet",
    "text": "pretty huge text will see what happens 4"
  },
  {
    "premium": false,
    "sender": "unnatural-error.testnet",
    "text": "pretty huge text will see what happens 3"
  },
  {
    "premium": false,
    "sender": "unnatural-error.testnet",
    "text": "pretty huge text will see what happens 2"
  },
  {
    "premium": false,
    "sender": "unnatural-error.testnet",
    "text": "pretty huge text will see what happens 1"
  }
]
```

Deployed V3:
```
handler error: [Function call returned an error: wasm execution failed with error: HostError(GuestPanic { panic_msg: "Cannot deserialize the contract state." })]
```

Migrated to V3:
```
   2: handler error: [Function call returned an error: wasm execution failed with error: HostError(GuestPanic { panic_msg: "Cannot deserialize element" })]
```

Reply:
Also, I'm not 100% confident about how StorageKey is generated, but I feel like it's not enough to replace a name, you need a new entry to be 100% safe. (e.g. change name to __RESERVED for example, and add new storage key. After you have cleared the entry, you might use this reserved key for other structs). Otherwise, you will overwrite the payments vector and might mess in a loop of that data.

@frol is the stated above correct?

Reply:
Given the fact that the contract uses `state_write` as well V3 structs are cached and write on drop only. It works just fine. I'm not sure where my error is coming from as I can no longer reproduce it.